sockeye/scoring.py

```diff
@@ -0,0 +1,270 @@
+        """
+        self.module.forward(batch, is_train=False)
+
+    def get_outputs(self):
+        return self.module.get_outputs()
+
+    def log_parameters(self):
+        """
+        Logs information about model parameters.
+        """
+        arg_params, aux_params = self.module.get_params()
+        total_parameters = 0
+        info = []  # type: List[str]
+        for name, array in sorted(arg_params.items()):
+            info.append("%s: %s" % (name, array.shape))
+            total_parameters += reduce(lambda x, y: x * y, array.shape)
+        logger.info("Model parameters: %s", ", ".join(info))
+        logger.info("Total # of parameters: %d", total_parameters)
+
+    def load_params_from_file(self, fname: str, allow_missing_params: bool = False):
+        """
+        Loads parameters from a file and sets the parameters of the underlying module and this model instance.
+
+        :param fname: File name to load parameters from.
+        :param allow_missing_params: If set, the given parameters are allowed to be a subset of the Module parameters.
+        """
+        super().load_params_from_file(fname)  # sets self.params & self.aux_params
+        self.module.set_params(arg_params=self.params,
+                               aux_params=self.aux_params,
+                               allow_missing=allow_missing_params)
+
+
+class Scorer:
+    def __init__(self,
+                 model: ScoringModel,
+                 source_vocabs: List[vocab.Vocab],
+                 target_vocab: vocab.Vocab,
+                 length_penalty: Optional[inference.LengthPenalty] = None) -> None:
+        self.source_vocab_inv = vocab.reverse_vocab(source_vocabs[0])
+        self.target_vocab_inv = vocab.reverse_vocab(target_vocab)
+        self.model = model
+        self.length_penalty = length_penalty
+
+        self.exclude_list = set([source_vocabs[0][C.BOS_SYMBOL], target_vocab[C.EOS_SYMBOL], C.PAD_ID])
+
+    def score(self,
+              score_iter,
+              score_type: str,
+              output: List[str]):
+
+        sentence_no = 0
+        for i, batch in enumerate(score_iter):
+            # data_io generates labels, too, which aren't needed in the computation graph
+            batch.provide_label = None
+            labels = batch.label[0].as_in_context(self.model.context[0])
+            batch.label = None
+            self.model.prepare_batch(batch)
+            self.model.run_forward(batch)
+            outputs = self.model.get_outputs()
+
+            batch_size, target_seq_len, _ = batch.provide_data[0][1]
+            outputs = mx.nd.reshape(data=outputs[0], shape=(-4, batch_size, target_seq_len, -2))
+
+            probs = mx.nd.pick(outputs, labels)
+            ones = mx.nd.ones_like(probs, ctx=self.model.context)
+            lengths = mx.nd.sum(labels != 0, axis=1) - 1
+
+            scores = mx.nd.log(mx.nd.where(labels != 0, probs, ones, ctx=self.model.context), ctx=self.model.context)
```

Comment:
Everything from getting the outputs up to here could be part of the computation graph, no need to do this imperatively.

Reply:
This is true, but having the `ScoringModel` return distributions could be useful (and in fact, is useful) for other applications.

A cleaner way to do this might be to parameterize `ScoringModel` with the desired output layer, i.e., one that returns either distributions or the desired scores. Do you prefer that?

Reply:
why not both by default? Graphs can have multiple outputs (grouped symbol).